### ISSUE TITLE
[member-ordering] variables-before-functions treat function properties as methods

### ISSUE THREAD

**Issue by mgol** [typescript-eslint/typescript-eslint#417]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/member-ordering": "error"
  }
}
```

```TS
const debounce = (fn: any, delay: any) => setTimeout(fn, delay);

class C {
  value = 42;
  returnValue = () => this.value;
  debouncedReturnValue = debounce(this.returnValue, 1000);
}
```

**Expected Result**

No error.

**Actual Result**

Error:
> Member debouncedReturnValue should be declared before all public instance method definitions

**Additional Info**

The problem is the `member-ordering` rule treats class fields defined as functions as methods when they are not methods. This creates problems - for methods, their order doesn't matter. For fields it sometimes does. In my example I have two fields with function values, the second one of them created not directly but by another function. I cannot change the order of those fields as the debounced returnValue function must already be defined when debouncedReturnValue is being defined.

This was previously reported to TSLint as https://github.com/palantir/tslint/issues/4156.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `1.5.0` |
| `@typescript-eslint/parser`        | `1.5.0` |
| `TypeScript`                       | `3.3.4000` |
| `ESLint`                           | `5.15.3` |
| `node`                             | `v8.15.1` |
| ~~`npm`~~                              | N/A |
| `yarn`                              | `1.15.2` |

**Comment by bradzacher** [comment]
We'd have to introduce type information into the rule in order to ensure that `debounce` is typed to return a function. Certainly doable.

The problem is that most people want to treat autobound methods as if they were normal methods, because otherwise there is an ordering discrepancy (esp in something like react, where you frequently mix and match autobound methods depending on if the function requires `this` or not).

Your use case is an edge case that isn't covered by the rule, and can't be covered.
Doubly so because you could turn on alphabetical sorting of fields, and then have a dependency which is broken by alphabetical sorting.

Note that you can work around this limitation by using a constructor.
Personally, I dislike directly referencing other fields within initialisers specifically because of this case. Also if you ever need to do logic more complex than can fit in an expression, it makes the diff a lot more noisy.

```TS
type debounce<T> = (fn: T, delay: any) => T;

class C {
  value = 42;
  debouncedReturnValue: () => void;
  returnValue = () => this.value;

  constructor() {
    this.debouncedReturnValue = debounce(this.returnValue, 1000);
  }
}
```

**Comment by bradzacher** [comment]
I'll close this issue as (IMO) your use case falls outside of the intended use of the rule.

We could do dependency tracking on functions and attempt to allow incorrect sorting if there are dependencies between field initialisers, but I think that is a lot of code for an edge case, and I would rather not have that level of complexity implemented.

**Comment by ashubham** [comment]
@bradzacher  I dont think this should be closed. This is not an edge case as it can be used to initialize members with an empty function for example.

```javascript
class MyComponent {
  onKeyDown: (evt) => any = (e) => undefined;
  onFocus: () => void = () => {};
  onBlur: () => void = () => {};

  private myField: number;
  constructor() {
  }
} 
```

**Comment by bradzacher** [comment]
you misunderstand the issue here.
This issue is because there is a dependency between the fields.

----

your issue is with treating "class properties with initial value set to a function expression" as methods instead of treating them as fields.
This is done on purpose. 99% of the time, the use case for class properties with function values is as a self-binding method definition.

Your use case of purely using a "class property with initial value set to a function expression" as storage space for a function reference is actually an edge case in the real world.

This will not be changed.

**Comment by ashubham** [comment]
Well I think the presence of configuration on lint rules is to cover the edge cases. This seems like a limitation of the tool, which is conveniently being dismissed as an edge case.
